core/src/tile/labels/label.cpp

```diff
@@ -76,61 +66,151 @@ void Label::updateScreenTransform(const glm::mat4& _mvp, const glm::vec2& _scree
             glm::vec2 t = glm::normalize(-p1p2);
-            
+            glm::vec2 tperp = glm::vec2(-t.y, t.x);
+
             float length = glm::length(p1p2);
-            
-            float exceedHeuristic = 40; // default heuristic : 40%
-            
-            if (m_width > length) {
-                float exceed = (1 - (length / m_width)) * 100;
-                m_visible &= exceed < exceedHeuristic;
+
+            float exceedHeuristic = 30; // default heuristic : 30%
+
+            if (m_dim.x > length) {
+                float exceed = (1 - (length / m_dim.x)) * 100;
+                if (exceed > exceedHeuristic) {
+                    return false;
+                }
             }
-            
-            screenPosition = (p1 + p2) / 2.0f;
-            
-            // translate by half the width
-            screenPosition += t * (m_width / 2);
+
+            screenPosition = (p1 + p2) * 0.5f + t * m_dim.x * 0.5f - tperp * (m_dim.y / 6);
 
             break;
         }
-    
-        default:
-            break;
     }
-    
+
     m_transform.m_screenPosition = screenPosition;
     m_transform.m_rotation = rot;
-    
-    m_outOfScreen = screenPosition.x > _screenSize.x || screenPosition.x < 0;
-    m_outOfScreen = m_outOfScreen || screenPosition.y > _screenSize.y || screenPosition.y < 0;
-    
-    m_dirty = true;
+
+    return true;
 }
 
 void Label::update(const glm::mat4& _mvp, const glm::vec2& _screenSize, float _dt) {
 
-    updateScreenTransform(_mvp, _screenSize);
-    updateBBoxes();
+    updateState(_mvp, _screenSize, _dt);
     
+    m_occlusionSolved = false;
+}
+
+bool Label::offViewport(const glm::vec2& _screenSize) {
+    const glm::vec2& screenPosition = m_transform.m_screenPosition;
+
+    bool outOfScreen = screenPosition.x > _screenSize.x || screenPosition.x < 0;
+    outOfScreen = outOfScreen || screenPosition.y > _screenSize.y || screenPosition.y < 0;
+
+    return outOfScreen;
+}
+
+void Label::setOcclusion(bool _occlusion) {
+    if (!canOcclude()) {
+        return;
+    }
+
+    m_occludedLastFrame = _occlusion;
+}
+
+bool Label::canOcclude() {
+    int occludeFlags = (State::VISIBLE | State::WAIT_OCC | State::FADING_IN);
+    return (occludeFlags & m_currentState) && !(m_type == Type::DEBUG);
+}
+
+void Label::occlusionSolved() {
+    m_occlusionSolved = true;
+}
+
+void Label::enterState(State _state, float _alpha) {
+    m_currentState = _state;
+    m_transform.m_alpha = CLAMP(_alpha, 0.0, 1.0);
+}
+
+void Label::updateState(const glm::mat4& _mvp, const glm::vec2& _screenSize, float _dt) {
+    if (m_currentState == State::SLEEP) {
+        // no-op state for now, when label-collision has less complexity, this state
+        // would lead to FADE_IN state if no collision occured
+        return;
+    }
+
+    bool occludedLastFrame = m_occludedLastFrame;
+    m_occludedLastFrame = false;
+
+    bool ruleSatisfied = updateScreenTransform(_mvp, _screenSize);
```

Comment:
The thing is there is not only the clipping test that is happening, also some rules that we could add later (now only one rule testing whether a line label fits inside the projected screen space anchor points).